src/lib/common/sol-platform-linux-common.c

```diff
@@ -263,3 +307,581 @@ sol_platform_impl_get_os_version(char **version)
+{
+    struct mntent mbuf = { 0 };
+    FILE *sol_mtab;
+    int err;
+
+    sol_mtab = setmntent(SOL_MTAB_FILE, "w+");
+    if (!sol_mtab) {
+        SOL_WRN("Unable to open "SOL_MTAB_FILE " file: %s",
+            sol_util_strerrora(-errno));
+        return -ENOENT;
+    }
+
+    mbuf.mnt_fsname = (char *)dev;
+    mbuf.mnt_dir = (char *)mpoint;
+    mbuf.mnt_type = (char *)fstype;
+    mbuf.mnt_opts = (char *)"";
+
+    err = addmntent(sol_mtab, &mbuf);
+    if (err != 0)
+        SOL_ERR("Could not add mnt entry - %s", sol_util_strerrora(-errno));
+
+    endmntent(sol_mtab);
+    return err;
+}
+
+static int
+run_mount(const char *dev, const char *mpoint, const char *fstype, unsigned long flags, const char *opts)
+{
+    int err;
+
+    err = sol_mtab_add_entry(dev, mpoint, fstype, flags);
+    SOL_INT_CHECK(err, < 0, err);
+
+    err = mount(dev, mpoint, fstype, flags, opts);
+    if (err != 0) {
+        SOL_ERR("Couldn't mount %s to %s - %s", dev, mpoint,
+            sol_util_strerrora(errno));
+        return -errno;
+    }
+
+    return err;
+}
+
+static void
+on_mount_fork(void *data)
+{
+    int err;
+    struct mount_data *mdata = data;
+
+    err = run_mount(mdata->dev, mdata->mpoint, mdata->fstype, mdata->flags, mdata->opts);
+    sol_platform_linux_fork_run_exit(err);
+}
+
+static void
+on_mount_fork_exit(void *data, uint64_t pid, int status)
+{
+    struct mount_data *mdata = data;
+
+    mdata->async_cb(mdata->dev, mdata->mpoint, mdata->data, pid, status);
+    free(mdata->opts);
+    free(mdata->fstype);
+    free(mdata->mpoint);
+    free(mdata->dev);
+    free(mdata);
+}
+
+int
+sol_platform_impl_mount(const char *dev, const char *mpoint, const char *fstype, unsigned long flags, const char *opts, void (*async_cb)(const char *dev, const char *mpoint, void *data, uint64_t pid, int status), void *data)
+{
+    struct mount_data *mdata;
+
+    SOL_NULL_CHECK(dev, -EINVAL);
+    SOL_NULL_CHECK(mpoint, -EINVAL);
+    SOL_NULL_CHECK(fstype, -EINVAL);
+
+    if (!async_cb)
+        return run_mount(dev, mpoint, fstype, flags, opts);
+
+    mdata = calloc(1, sizeof(struct mount_data));
+    SOL_NULL_CHECK(mdata, -ENOMEM);
+
+    mdata->data = data;
+    mdata->dev = strdup(dev);
+    mdata->mpoint = strdup(mpoint);
+    mdata->fstype = strdup(fstype);
+
+    if (opts)
+        mdata->opts = strdup(opts);
+    
+    mdata->flags = flags;
+    mdata->async_cb = async_cb;
+
+    sol_platform_linux_fork_run(on_mount_fork, on_mount_fork_exit, mdata);
+    return 0;
+}
+
+static void
+sol_uevent_event_dispatch(struct uevent_context *ctx, struct sol_uevent *uevent)
+{
+    uint16_t idx;
+    struct uevent_callback *cb;
+
+    SOL_VECTOR_FOREACH_IDX (&ctx->callbacks, cb, idx) {
+        if ((!cb->action || (cb->action && sol_str_slice_str_eq(uevent->action, cb->action))) &&
+            (!cb->subsystem || (cb->subsystem && sol_str_slice_str_eq(uevent->subsystem, cb->subsystem)))) {
+            cb->uevent_cb((void *)cb->cb_data, uevent);
```

Comment:
you need to be careful here, if `uevent_cb` changes `ctx->callbacks` (ie: remove a callback handler), it will crash.

See other parts of our code where we keep a pending list + walking counter.